File: FBSimulatorControl/Notifiers/FBCoreSimulatorNotifier.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBCoreSimulatorNotifier.h"

#import <CoreSimulator/SimDevice.h>
#import <CoreSimulator/SimDeviceNotificationManager.h>
#import <CoreSimulator/SimDeviceNotifier-Protocol.h>
#import <CoreSimulator/SimDeviceSet.h>

#import "FBSimulator.h"
#import "FBSimulatorSet.h"

@interface FBCoreSimulatorNotifier ()

@property (nonatomic, readonly, assign) NSUInteger handle;
@property (nonatomic, readonly, strong) id notifier;

@end

@implementation FBCoreSimulatorNotifier

#pragma mark Public

+ (instancetype)notifierForSimDevice:(SimDevice *)simDevice queue:(dispatch_queue_t)queue block:(void (^)(NSDictionary<NSString *, id> *info))block
{
  id<NSObject, SimDeviceNotifier> notifier = simDevice.notificationManager;
  return [[self alloc] initWithNotifier:notifier queue:queue block:block];
}

+ (FBFuture<NSNull *> *)resolveLeavesState:(FBiOSTargetState)state forSimDevice:(SimDevice *)device
{
  FBMutableFuture<NSNull *> *future = FBMutableFuture.future;
  dispatch_queue_t queue = dispatch_queue_create("com.facebook.fbsimulatorcontrol.resolve_state", DISPATCH_QUEUE_SERIAL);
  FBCoreSimulatorNotifier *notifier = [self notifierForSimDevice:device queue:queue block:^(NSDictionary<NSString *, id> *info) {
    if (![info[@"notification"] isEqualToString:@"device_state"]) {
      return;
    }
    NSNumber *newStateNumber = info[@"new_state"];
    if (!newStateNumber) {
      return;
    }
    FBiOSTargetState newState = newStateNumber.unsignedIntegerValue;
    if (newState == state) {
      return;
    }
    [future resolveWithResult:NSNull.null];
  }];
  return [future
   onQueue:queue notifyOfCompletion:^(id _) {
     [notifier terminate];
   }];
}

- (void)terminate
{
  [self.notifier unregisterNotificationHandler:self.handle error:nil];
}

#pragma mark Private

+ (instancetype)notifierForSet:(FBSimulatorSet *)set queue:(dispatch_queue_t)queue block:(void (^)(NSDictionary<NSString *, id> *info))block
{
  id<NSObject, SimDeviceNotifier> notifier = set.deviceSet.notificationManager;
  return [[self alloc] initWithNotifier:notifier queue:queue block:block];
}

- (instancetype)initWithNotifier:(id<NSObject, SimDeviceNotifier>)notifier queue:(dispatch_queue_t)queue block:(void (^)(NSDictionary<NSString *, id> *info))block
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _notifier = notifier;
  _handle = [FBCoreSimulatorNotifier registerNotificationHandler:notifier queue:queue handler:block];

  return self;
}

+ (NSUInteger)registerNotificationHandler:(id<NSObject, SimDeviceNotifier>)notifier queue:(dispatch_queue_t)queue handler:(void (^)(NSDictionary<NSString *, id> *))handler
{
  if ([notifier respondsToSelector:@selector(registerNotificationHandlerOnQueue:handler:)]) {
    return [notifier registerNotificationHandlerOnQueue:queue handler:handler];
  } else {
    return [notifier registerNotificationHandler:^(NSDictionary *info) {
      dispatch_async(queue, ^{
        handler(info);
      });
    }];
  }
}

@end

File: FBSimulatorControl/Notifiers/FBCoreSimulatorNotifier.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@_implementationOnly import CoreSimulator
@preconcurrency import FBControlCore
import Foundation

@objc(FBCoreSimulatorNotifier)
public final class FBCoreSimulatorNotifier: NSObject {

  // MARK: - Properties

  private let handle: UInt64
  private let notifierObj: AnyObject? // SimDeviceNotifier (nil in test doubles)

  // MARK: - Public

  @objc(notifierForSimDevice:queue:block:)
  public class func notifier(for simDevice: SimDevice, queue: DispatchQueue, block: @escaping @Sendable ([String: Any]) -> Void) -> FBCoreSimulatorNotifier {
    let notifier = simDevice.notificationManager as AnyObject?
    return FBCoreSimulatorNotifier(notifier: notifier, queue: queue, block: block)
  }

  @objc(resolveLeavesState:forSimDevice:)
  public class func resolveLeavesState(_ state: FBiOSTargetState, for device: SimDevice) -> FBFuture<NSNull> {
    let future = FBMutableFuture<NSNull>()
    let queue = DispatchQueue(label: "com.facebook.fbsimulatorcontrol.resolve_state")
    nonisolated(unsafe) let futureRef = future
    let notifier = self.notifier(for: device, queue: queue) { info in
      guard let notification = info["notification"] as? String, notification == "device_state" else {
        return
      }
      guard let newStateNumber = info["new_state"] as? NSNumber else {
        return
      }
      let newState = FBiOSTargetState(rawValue: newStateNumber.uintValue)
      if newState == state {
        return
      }
      futureRef.resolve(withResult: NSNull())
    }
    return
      unsafeBitCast(future, to: FBFuture<NSNull>.self)
      .onQueue(
        queue,
        notifyOfCompletion: { (_: Any) in
          notifier.terminate()
        })
  }

  @objc
  public func terminate() {
    guard let notifier = notifierObj else { return }
    (notifier as! SimDeviceNotifier).unregisterNotificationHandler(handle, error: nil)
  }

  // MARK: - Internal

  @objc(notifierForSet:queue:block:)
  class func notifier(for set: FBSimulatorSet, queue: DispatchQueue, block: @escaping @Sendable ([String: Any]) -> Void) -> FBCoreSimulatorNotifier {
    // notificationManager may be nil in test doubles (ObjC nil messaging returns nil).
    let notifier = (set.deviceSet as AnyObject).notificationManager as AnyObject?
    return FBCoreSimulatorNotifier(notifier: notifier, queue: queue, block: block)
  }

  // MARK: - Private

  private init(notifier: AnyObject?, queue: DispatchQueue, block: @escaping @Sendable ([String: Any]) -> Void) {
    self.notifierObj = notifier
    // nil for test doubles; mirror ObjC nil-messaging with a 0 handle.
    guard let notifier = notifier as? SimDeviceNotifier else {
      self.handle = 0
      super.init()
      return
    }
    self.handle = notifier.registerNotificationHandler(on: queue) { info in
      block((info as? [String: Any]) ?? [:])
    }
    super.init()
  }
}
